core/test/utils/matrix_generator.hpp

```diff
@@ -82,22 +84,48 @@ matrix_data<ValueType, IndexType> generate_random_matrix_data(
 
     matrix_data<ValueType, IndexType> data{gko::dim<2>{num_rows, num_cols}, {}};
 
-    std::vector<size_type> col_idx(num_cols);
-    std::iota(begin(col_idx), end(col_idx), size_type(0));
+    std::vector<bool> present_cols(num_cols);
 
-    for (size_type row = 0; row < num_rows; ++row) {
+    for (IndexType row = 0; row < num_rows; ++row) {
         // randomly generate number of nonzeros in this row
-        auto nnz_in_row = static_cast<size_type>(nonzero_dist(engine));
-        nnz_in_row = std::max(size_type(0), std::min(nnz_in_row, num_cols));
-        // select a subset of `nnz_in_row` column indexes, and fill these
-        // locations with random values
-        std::shuffle(begin(col_idx), end(col_idx), engine);
-        std::for_each(
-            begin(col_idx), begin(col_idx) + nnz_in_row, [&](size_type col) {
-                data.nonzeros.emplace_back(
-                    row, col,
-                    detail::get_rand_value<ValueType>(value_dist, engine));
-            });
+        const auto nnz_in_row = std::max(
+            size_type(0),
+            std::min(static_cast<size_type>(nonzero_dist(engine)), num_cols));
+        std::uniform_int_distribution<IndexType> col_dist{
+            0, static_cast<IndexType>(num_cols) - 1};
+        if (nnz_in_row > num_cols / 2) {
+            present_cols.assign(num_cols, true);
+            // remove num_cols - nnz_in_row entries from present_cols
+            size_type count = num_cols;
+            while (count > nnz_in_row) {
+                const auto new_col = col_dist(engine);
+                if (present_cols[new_col]) {
+                    present_cols[new_col] = false;
+                    count--;
+                }
+            }
+            for (IndexType col = 0; col < num_cols; col++) {
+                if (present_cols[col]) {
+                    data.nonzeros.emplace_back(
+                        row, col,
+                        detail::get_rand_value<ValueType>(value_dist, engine));
+                }
+            }
+        } else {
+            // add nnz_in_row entries to present_cols
```

Comment:
If I interpret the constructor right, then no reservation has been done.
Since we do know that we will have exactly `nnz_in_row` elements, then we can safely reserve that amount

Reply:
I somehow hyperfocused on the inner loop and forgot the outer loop. Either an average would be required, but it's generally hard to guess I think.